### DISCUSSION TITLE
Attach interface to class

### DISCUSSION THREAD

**Discussion by bvbellomo** [dotnet/csharplang#5300] [Language Ideas]
Suppose my code consumes a class I did not write and cannot modify.  It may fit a pattern of several other similar classes, and I may have code that works with other similar classes.  My code cannot easily work with this class - I might be able to derive a new class or write a wrapper, but this is awkward and not always possible.

In Python, I have duck typing.  I don't have to worry about declaring whether a class supports an interface.  In C#, we do not have duck typing, and require interfaces and classes to be explicitly declared.  I understand the reason for that.

However, we are forced to declare which interfaces a class supports in the class declaration itself -e.g. 
public class Widget : IWidget

This makes it impossible to declare an IWidget interface for our code to consume if Widget is defined elsewhere (such as the .NET Framework) and we are writing the IWidget interface and code to consume it, even if the Widget class already has the methods and properties it needs for our interface.

It is syntactically more challenging, but we could allow the declaration of an interface to include the classes that implement it, e.g.
public interface IWidget(_supports_:System.Microsoft.SealedClasses.Widget)

or we could allow declaring that a class supports an interface outside the declaration of the class and the declaration of the interface.  e.g.
public class System.Microsoft.SealedClasses.Widget : IWidget;

This would allow a lot more flexibility working with externally defined classes.

**Comment by 333fred** [comment]
See https://github.com/dotnet/csharplang/issues/1711.
